### DISCUSSION TITLE
Apply buoyancy impulse offset

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#76] [Ideas]
Was playing with the apply buoyancy impulse example and noticed the objects keep floating in mid air, if they are even slightly touching the water sensor. I am thinking of a scenario, where an object floats on a river, then falls down reaching a waterfall. What would be the way to restrict object to float in mid air, when it reaches the waterfall/river edge? Maybe making the sensor width shorter than the visual mesh? Or applying buoyancy impulse at an offset (although, not sure we have API for that)?

**Comment by PhoenixIllusion** [comment]
Buoyancy interacts with an object using a planar formula. The water position and normal in that equation are used to construct a plane used to slice the volume and calculate the amount of the object volume and then apply a force counter to gravity scaled against he intersected body. https://github.com/jrouwe/JoltPhysics/blob/master/Jolt/Physics/Body/Body.cpp#L186

The concept of fluids with this equation is not concretely related to the sensor areas, other than the area saying "the fluid exists in this general area", and applying what you are describing will likely map more to a conceptual-formula of water vs discrete XYZ/bounding-box/sensor.

In the concept of a waterfall, the most direct way to apply what you are saying is to create a formula, singular or piece-wise, that can take a given XYZ+extent of an object and derive a given plane and directional velocity.

I had thought of implementing this in my example, but had not yet done it. In the visual example, in 2D, given a 'ball' is in the blue-hatched bounding box, you would need to take the size of the ball and the X/Y coordinate to derive the proper surface normal and fluid velocity (represented by the black and blue arrows along the curve). This can be converted into a simple curved equation.

Given the ball technically is larger than any given slice of the complex curve, and we can only give the buoyancy impulse equation an infinite plane, you will likely would just calculate an estimate unless you want to derive an equivalent combination of [position, velocity, normal].

In the image below, I probably would not go with a full equation and just split it into a 3-part piecewise lookup. The water on the very top-right of the waterfall, the water on the bottom-left of the waterfall, and then the transition point with a 1x diagonal surface.

```
if( ball_in_waterfall) {
    const position_in_waterfall = get_rough_position(ball_xyz, waterfall_extent);
    switch(position_in_waterfall) {
         case Waterfall.ON_TOP:
           fluid_surface_position = ///
           fluid_surface_normal = ///
           fluid_velocity = ///
           break;
         case Waterfall.TRANSITIONING:
         case Waterfall.FALLING:
    }
}
```
![2023-12-14-fluid-buoyancy](https://github.com/jrouwe/JoltPhysics.js/assets/756488/506e4e9b-2522-4ed1-9e30-6b3a12395b7d)

**Reply by PhoenixIllusion** [reply]
I believe your concept of offset would be applying one or two additional planar slices in the equation outside of the built-in one and further cutting down the buoyancy by additional fractions, ex: if a bodies bounding box or something is only 10% in an area sensor, keep the position/normal up but only apply 10% buoyancy. This could partially approximate some things, but doesn't quite fully work to properly handle edge cases where water only applies forces to part of an object (center of buoyancy in the source),since buoyancy impulse only influences parts of an body actually in the water, ex a water wheel should turn when the outer ring is in moving water. Applying calculated planar descriptions of the fluid surface should mostly solve for these.

Something to take into account is that if a buoyancy surface lookup was theoretically calculated correctly, you could apply it to every dynamic body anywhere in the world every single physics step and the impulse equation would just say "the user is on the opposite side of my position/normal plane, do nothing. Intersected volume is 0.

One of the bigger challenges in this lookup is describing surfaces that are transitioning between wildly different planes or smaller than your colliding volume, otherwise it is almost just "describe as a convex hull and use nearest faces and average any intersected hull faces" since hulls are technically spatially carved volumes composed of plane intersections, so have a lot of the data readily available.

I was thinking through how to simplify or solve some of the challenging cases by applying the impulse with a summation of slices, a crude integration of forces on a body to handle something like a log falling lengthwise off the cliff and having support only on one parts of the log as it transitions over the waterfall. Properly having the center-of-buoyancy calculate to the right locations on a body per slice of force is challenging, but I may look into it.

**Reply by PhoenixIllusion** [reply]
It seems that technically the mentally clearer solution to my problem is that if a desired floating body gets too large, break the body into parts and glue them together with constraints. That way the same compound body going through the volume will naturally have the waters field equation hit each chunk of the sub-body with the appropriate forces and Center of Buoyancy? That also applies if the water direction changes as you are going around a corner and need the front of a boat to get pushed one direct and the back to be pushed another direction

**Reply by LeXXik** [reply]
Oh, that's actually a neat idea! Will give it a try.

**Comment by LeXXik** [comment]
Ah, that is a good idea @PhoenixIllusion. Thank you for the illustration, which makes it clear. I find your answers to be always very educational :) I will give it a try.
